packages/protocol/src/vaults/borrowing/BorrowingVaultUpgradeable.sol

```diff
@@ -548,6 +526,45 @@ contract BorrowingVaultUpgradeable is BaseVaultUpgradeable {
     }
   }
 
+  /**
+   * @dev Function to handle common flow for `payback(...)` and `burnDebt(...)`
+   * It returns the updated `debt` and `shares` values if applicable.
+   *
+   * @param debt or borrowed amount of debt asset
+   * @param shares amount of `debtShares`
+   * @param owner to whom `debtShares` will bet burned
+   * @param caller msg.sender
+   */
+  function _paybackInternal(
+    uint256 debt,
+    uint256 shares,
+    address owner,
+    address caller
+  )
+    internal
+    returns (uint256 debt_, uint256 shares_)
+  {
+    // This local var helps save gas not having to call provider balances again
+    // and is multiplied by debtAsset decimals to mantain precision.
+    uint256 savedDebtSharesRatio = shares > 0 ? debt.mulDiv(10 ** _debtDecimals, shares) : 0;
+
+    uint256 remainder;
+    // `debt`, `shares`are updated if passing more than max amount for `owner`'s debt.
+    (debt_, shares_, remainder) = _paybackChecks(owner, debt, shares, savedDebtSharesRatio);
+
+    _payback(caller, owner, debt_, shares_);
+
+    if (remainder > 0) {
+      /**
+       * @devSince the `_payback(...) only pulls (erc20) that is needed to payback
+       * maxAmount, this logic handles excess amount `remainder` by pulling from
+       * `msg.sender and returning to the `owner` the `remainder`.
+       */
+      _debtAsset.safeTransferFrom(caller, address(this), remainder);
+      _debtAsset.safeTransfer(owner, remainder);
```

Comment:
Why not do this with one call?
`_debtAsset.safeTransferFrom(caller, owner, remainder);`

A similar note for "withdraw".